File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-1.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//     pub val: i32,
//     pub next: Option<Box<ListNode>>,
// }
//
// impl ListNode {
//     #[inline]
//     pub fn new(val: i32) -> Self {
//         ListNode { next: None, val }
//     }
// }

impl Solution {
    pub fn merge_two_lists(mut list1: Option<Box<ListNode>>, mut list2: Option<Box<ListNode>>) -> Option<Box<ListNode>> {

        let mut dummy = Box::new(ListNode::new(0));

        let mut node = &mut dummy;

        while let (Some(l1), Some(l2)) = (list1.as_ref(), list2.as_ref()) {
            if l1.val < l2.val {
                node.next = list1;
                node = node.next.as_mut().unwrap();
                list1 = node.next.take();
            } else {
                node.next = list2;
                node = node.next.as_mut().unwrap();
                list2 = node.next.take();
            }
        }

        node.next = if list1.is_some() { list1 } else { list2 };

        dummy.next
    }
}

File: Data Structures & Algorithms/add-two-numbers/submission-1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode dummy;
        auto curr = &dummy;
        auto left_over{0};
        while(l1 || l2) {
            auto l1_val{0};
            if (l1) {
                l1_val = l1->val;
                l1 = l1->next;
            }
            auto l2_val{0};
            if (l2) {
                l2_val = l2->val;
                l2 = l2->next;
            }
            auto new_val{l1_val + l2_val + left_over};
            if(new_val < 10) {
                curr->next = new ListNode(new_val);
                curr = curr->next;
                left_over = 0;
            } else {
                auto left = new_val - 10;
                auto right{1};
                curr->next = new ListNode(left);
                curr = curr->next;
                left_over = 1;
            }
        }

        if(left_over != 0) {
            curr->next = new ListNode(1);
        }

        return dummy.next;
    }
};
